File: hot100/03_Sliding_Window/0076_Minimum_Window_Substring/solution.cpp
#include <bits/stdc++.h>
using namespace std;

#define GRN "\033[32m"
#define RED "\033[31m"
#define BLD "\033[1m"
#define RST "\033[0m"

// ─── Solution ────────────────────────────────────────────────────────────────

class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> tCnt;
        for (auto &c : t) {
            tCnt[c]++;
        }
        auto check = [&] (const unordered_map<char, int> &cnt) {
            for (auto &[k, v] : cnt) {
                if (v > 0) return false;
            }
            return true;
        };

        int l = 0, r = 0;
        int n = s.size();

        // move r to legal
        while (r < n && !check(tCnt)) {
            tCnt[s[r]]--;
            r++; // add s[r] in it
        }

        if (r == n && !check(tCnt)) return ""; // Not Found

        // move l
        while (l < n && tCnt[s[l]] + 1 <= 0) {
            tCnt[s[l]]++;
            l++;
        }
        int ans = r-l;
        int ansl = l;
        int ansr = r;

        while (r < n) {
            // add r
            tCnt[s[r]]--;
            r++;

            // move l
            while (l < n && tCnt[s[l]] + 1 <= 0) {
                tCnt[s[l]]++;
                l++;
            }
            if (r - l < ans) {
                ans = r-l;
                ansl = l;
                ansr = r;
            }

        }
        return s.substr(ansl, ansr - ansl);

    }
};

// ─── Helpers ─────────────────────────────────────────────────────────────────

auto toString = [](string v) { return "\"" + v + "\""; };
auto eq       = [](string a, string b) { return a == b; };

// ─── Test Harness ─────────────────────────────────────────────────────────────

int main() {
    Solution sol;
    int passed = 0, total = 0;

    auto check = [&](string got, string want, const string& label) {
        total++;
        if (eq(got, want)) {
            printf(GRN "✓" RST "  %-36s →  %s\n", label.c_str(), toString(got).c_str());
            passed++;
        } else {
            printf(RED "✗" RST "  " BLD "%-36s" RST "    want=%s  got=%s\n",
                   label.c_str(), toString(want).c_str(), toString(got).c_str());
        }
    };

    // All problem examples + ≥2 edge cases
    check(sol.minWindow("ADOBECODEBANC", "ABC"), "BANC", "example1: ADOBECODEBANC/ABC");
    check(sol.minWindow("a",              "a"),  "a",    "example2: a/a");
    check(sol.minWindow("a",              "aa"), "",     "example3: a/aa");
    check(sol.minWindow("b",              "b"),  "b",    "edge: single match");
    check(sol.minWindow("abc",            "cba"), "abc", "edge: full string");
    check(sol.minWindow("xyzabc",         "abc"), "abc", "edge: suffix match");
    check(sol.minWindow("AA",             "AA"), "AA",   "edge: duplicate required chars");

    printf("\n%s\n", string(44, '-').c_str());
    if (passed == total)
        printf(GRN "%d / %d passed\n" RST, passed, total);
    else
        printf(RED "%d / %d passed\n" RST, passed, total);
    return (passed == total) ? 0 : 1;
}

File: hot100/03_Sliding_Window/0076_Minimum_Window_Substring/solution.go
package main

import "fmt"

const (
	grn = "\033[32m"
	red = "\033[31m"
	bld = "\033[1m"
	rst = "\033[0m"
)

// ─── Solution ────────────────────────────────────────────────────────────────

func minWindow(s string, t string) string {
	// TODO: implement
	return ""
}

// ─── Test Harness ─────────────────────────────────────────────────────────────

func main() {
	passed, total := 0, 0

	check := func(got, want string, label string) {
		total++
		if got == want {
			fmt.Printf(grn+"✓"+rst+"  %-36s →  %q\n", label, got)
			passed++
		} else {
			fmt.Printf(red+"✗"+rst+"  "+bld+"%-36s"+rst+"    want=%q  got=%q\n", label, want, got)
		}
	}

	// All problem examples + ≥2 edge cases
	check(minWindow("ADOBECODEBANC", "ABC"), "BANC", "example1: ADOBECODEBANC/ABC")
	check(minWindow("a",              "a"),  "a",    "example2: a/a")
	check(minWindow("a",              "aa"), "",     "example3: a/aa")
	check(minWindow("b",              "b"),  "b",    "edge: single match")
	check(minWindow("abc",            "cba"), "abc", "edge: full string")
	check(minWindow("xyzabc",         "abc"), "abc", "edge: suffix match")
	check(minWindow("AA",             "AA"), "AA",   "edge: duplicate required chars")

	fmt.Printf("\n%s\n", "--------------------------------------------")
	if passed == total {
		fmt.Printf(grn+"%d / %d passed\n"+rst, passed, total)
	} else {
		fmt.Printf(red+"%d / %d passed\n"+rst, passed, total)
	}
}
